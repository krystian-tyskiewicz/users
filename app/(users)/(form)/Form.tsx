'use client'

import { styled } from 'styled-components'
import { useForm, Controller, SubmitHandler } from 'react-hook-form'

import {
  Input,
  ButtonsGroup,
  FormButton,
  FormRow,
  FormInputError,
} from '@/app/components'
import { UserInput } from '@/app/types'

const Form = styled.form``

type Props = {
  onSubmit: SubmitHandler<UserInput>
  defaultValues?: UserInput
}

const UserForm = ({ onSubmit, defaultValues }: Props) => {
  const { control, handleSubmit, formState: { errors } } = useForm({
    defaultValues: defaultValues || {
      name: '',
      username: '',
      email: '',
      city: '',
    },
  })

  return (
    <Form onSubmit={handleSubmit(onSubmit)}>
      <FormRow>
        <Controller
          name="name"
          control={control}
          rules={{
            required: true,
          }}
          render={({ field }) => <Input placeholder="Name*" {...field} />}
        />
      </FormRow>
      <FormRow>
        <Controller
          name="username"
          control={control}
          rules={{
            required: true,
          }}
          render={({ field }) => <Input placeholder="Username*" {...field} />}
        />
      </FormRow>
      <FormRow>
        <Controller
          name="email"
          control={control}
          rules={{
            required: true,
            pattern: {
              value: /\S+@\S+\.\S+/,
              message: "Entered value does not match email format"
            },
          }}
          render={({ field }) => <Input placeholder="Email*" {...field} />}
        />
        {errors.email && <FormInputError>{errors.email?.message}</FormInputError>}
      </FormRow>
      <FormRow>
        <Controller
          name="city"
          control={control}
          render={({ field }) => <Input placeholder="City" {...field} />}
        />
      </FormRow>
      <FormRow>
      <ButtonsGroup>
        <FormButton danger href="/home">Cancel</FormButton>
        <FormButton success type="submit">Submit</FormButton>
      </ButtonsGroup>
      </FormRow>
    </Form>
  )
}

export default UserForm
