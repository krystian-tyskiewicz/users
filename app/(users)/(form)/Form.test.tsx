import { render } from '@testing-library/react'
import Form from './Form'

jest.mock('next/navigation', () => ({
  useRouter: jest.fn(),
}))

it('renders correctly', () => {
  const props = {
    onSubmit: jest.fn(),
  }
  const { container } = render(<Form {...props} />)
  expect(container).toMatchSnapshot()
})
