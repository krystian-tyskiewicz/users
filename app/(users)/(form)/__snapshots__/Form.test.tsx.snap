// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <form
    class="sc-ksdynI"
  >
    <div
      class="sc-eCIlHX kAWkuI"
    >
      <input
        class="sc-gKclGN kytkRg"
        name="name"
        placeholder="Name*"
        value=""
      />
    </div>
    <div
      class="sc-eCIlHX kAWkuI"
    >
      <input
        class="sc-gKclGN kytkRg"
        name="username"
        placeholder="Username*"
        value=""
      />
    </div>
    <div
      class="sc-eCIlHX kAWkuI"
    >
      <input
        class="sc-gKclGN kytkRg"
        name="email"
        placeholder="Email*"
        value=""
      />
    </div>
    <div
      class="sc-eCIlHX kAWkuI"
    >
      <input
        class="sc-gKclGN kytkRg"
        name="city"
        placeholder="City"
        value=""
      />
    </div>
    <div
      class="sc-eCIlHX kAWkuI"
    >
      <div
        class="sc-gsDKeI ejuFsk"
      >
        <button
          class="sc-bdvumH ynMys sc-dkPulx lgyEdo"
          type="button"
        >
          Cancel
        </button>
        <button
          class="sc-bdvumH ftCnfd sc-dkPulx lgyEdo"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </form>
</div>
`;
